### PR TITLE
Exercise-018 Windowing System

### PR DESCRIPTION
**Pull request by Dreamless**

### CODE REVIEW

File: javascript/windowing-system/windowing-system.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    this.screenSize = new Size(800, 600);
+    this.size = new Size();
+    this.position = new Position();
+  }
+
+  resize(newSize) {
+    let sizeW = newSize.width > 1 ? newSize.width : 1;
+    let sizeH = newSize.height > 1 ? newSize.height : 1;
+    const shiftSizeW = newSize.width + sizeW;
+    const shiftSizeH = newSize.height + sizeH;
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.js:37]: ?!

File: javascript/windowing-system/windowing-system.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    this.size.resize(sizeW, sizeH);
+  }
+
+  move(newPosition) {
+    let posX = newPosition.x > 0 ? newPosition.x : 0;
+    let posY = newPosition.y > 0 ? newPosition.y : 0;
+    const shiftWidth = this.size.width + posX;
+    const shiftHeight = this.size.height + posY;
+
+    if(shiftWidth > this.screenSize.width) {
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.js:56]: Use IDE formatter

File: javascript/windowing-system/windowing-system.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    this.size.resize(sizeW, sizeH);
+  }
+
+  move(newPosition) {
+    let posX = newPosition.x > 0 ? newPosition.x : 0;
+    let posY = newPosition.y > 0 ? newPosition.y : 0;
+    const shiftWidth = this.size.width + posX;
+    const shiftHeight = this.size.height + posY;
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.js:54]: Misleading names, those are not width and height

File: javascript/windowing-system/windowing-system.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    this.position.move(posX, posY);
+  }
+}
+
+
+export function changeWindow(updatedWindow) {
+  updatedWindow.position.move(100, 150);
+  updatedWindow.size.resize(400, 300)
+  return updatedWindow;
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.js:72]: Add a test like this:
```
Screen size:     800x600 (default)
Window size:     800x600 (same as screen)
Window position: 0x0

changeWindow is called
Verify size and position
```

**Review by frozenspider — CHANGES_REQUESTED** [review]

File: javascript/windowing-system/windowing-system.js (hunk 4 above)
Comment (Dreamless) [reply javascript/windowing-system/windowing-system.js:72]: please look, I'm not sure that the test is correct

File: javascript/windowing-system/windowing-system.spec.js [hunk 5]
```diff
@@ -153,6 +153,18 @@ describe('changeWindow', () => {
...
+  test('verify window size and position', () => {
+    const programWindow = new ProgramWindow();
+    const newSize = new Size(800, 600);
+    programWindow.resize(newSize);
+    const newPosition = new Position(0, 0);
+    programWindow.move(newPosition);
+    const updatedWindow = changeWindow(programWindow);
+
+    expect(updatedWindow.position.x).toBe(100);
+    expect(updatedWindow.position.y).toBe(150);
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.spec.js:165]: You call this test "verify window size and position", but you're not even checking size.

File: javascript/windowing-system/windowing-system.js [hunk 6]
```diff
@@ -31,33 +31,33 @@ export class ProgramWindow {
...
-    this.size.resize(sizeW, sizeH);
+    this.size.resize(newWidth, newHeight);
   }
 
   move(newPosition) {
     let posX = newPosition.x > 0 ? newPosition.x : 0;
     let posY = newPosition.y > 0 ? newPosition.y : 0;
-    const shiftWidth = this.size.width + posX;
-    const shiftHeight = this.size.height + posY;
+    const posBasedOnSizeX = this.size.width + posX;
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.js:53]: What is the name of that position? Can you come up with a better name?

File: javascript/windowing-system/windowing-system.js [hunk 7]
```diff
@@ -31,33 +31,33 @@ export class ProgramWindow {
   }
 
   resize(newSize) {
-    let sizeW = newSize.width > 1 ? newSize.width : 1;
-    let sizeH = newSize.height > 1 ? newSize.height : 1;
-    const shiftSizeW = newSize.width + sizeW;
-    const shiftSizeH = newSize.height + sizeH;
+    let newWidth = newSize.width > 1 ? newSize.width : 1;
+    let newHeight = newSize.height > 1 ? newSize.height : 1;
+    const maxWidth = newSize.width + newWidth;
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.js:36]: This code is still broken by design.
```js
describe('move', () => {

  // ...

  test('fix me', () => {
    const programWindow = new ProgramWindow();
    const newPosition = new Position(programWindow.screenSize.width - 100, programWindow.screenSize.height - 100);

    programWindow.move(newPosition);
    const newSize = new Size(200, 200);
    programWindow.resize(newSize);

    expect(programWindow.size.width).toBe(100);
    expect(programWindow.size.height).toBe(100);
  });
});
```

Besides, this is NOT maximum width. Start thinking about what values mean.

**Review by frozenspider — CHANGES_REQUESTED** [review]
Why is the exercise named `windowing-system` without number?

File: javascript/windowing-system/windowing-system.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  constructor() {
+    this.screenSize = new Size(800, 600);
+    this.size = new Size();
+    this.position = new Position();
+  }
+
+  resize(newSize) {
+    let newWidth = newSize.width > 1 ? newSize.width : 1;
+    let newHeight = newSize.height > 1 ? newSize.height : 1;
+    const newWindowWidth = this.position.x + newWidth;
```
Comment (frozenspider) [inline javascript/windowing-system/windowing-system.js:36]: Naming is still misleading, in both methods

**Review by frozenspider — APPROVED** [review]

**Pull request merged**